docs/petr4spec/Petr4-spec.mdk

```diff
@@ -142,509 +149,2979 @@ p4grammar {
+|   | $\vert$ | $\block \stmts$ | $\qquad \text{sequencing}$ |
+|   | $\vert$ | $\exit$ | $\qquad \text{exit}$ |
+|   | $\vert$ | $\noop$ | $\qquad \text{noop}$ |
+|   | $\vert$ | $\return \exp$ | $\qquad \text{return}$ |
+|   | $\vert$ | $\retNothing$ | $\qquad \text{return nothing}$ |
+|   | $\vert$ | $\switch \exp \switchCases$ | $\qquad \text{switch}$ |
+|   | $\vert$ | $\dcl$ | $\qquad \text{declaration}$ |
+
+**Parser states:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\nxt$ | $::=$ | $\str$ | $\qquad \text{next state}$ |
+| $\pmatch$ | $::=$ | $\defMatch$ | $\qquad \text{default match}$ |
+|   | $\vert$ | $\dontcareMatch$ | $\qquad \text{don't care match}$ |
+|   | $\vert$ | $\exp$ | $\qquad \text{expression match}$ |
+| $\case$ | $::=$ | $\expandCase \pmatches \nxt$ | $\qquad \text{match cases for the next state}$ |
+| $\parserTran$ | $::=$ | $\nxt$ | $\qquad \text{next state transition}$ |
+|   | $\vert$ | $\slctTran \exp \pcases$ | $\qquad \text{select next state transition}$ |
+| $\state$ | $::=$ | $\stateDef \var \stmts \parserTran$ | $\qquad \text{states}$ |
+
+**Table properties:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\action$ | $::=$ | $\actionDef \var \arg$ | $\qquad \text{actions}$ |
+| $\prop$ | $::=$ | $\key \exp \var$ | $\qquad \text{keys}$ |
+|   | $\vert$ | $\action$ | $\qquad \text{actions}$ |
+|   | $\vert$ | $\entry \pmatch \action$ | $\qquad \text{entries}$ |
+|   | $\vert$ | $\custom \bool \var \exp$ | $\qquad \text{custom property}$ |
+
+**Methods:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\extMethod$ | $::=$ | $\constructor \var \prmDef$ | $\qquad \text{constructors}$ |
+|   | $\vert$ | $\abstractMethod \baseTyp \var \typVar \prmDef$ | $\qquad \text{abstract methods}$ |
+|   | $\vert$ | $\methodDef \baseTyp \var \typVar \prmDef$ | $\qquad \text{methods}$ |
+
+**Declarations:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\dcl$ | $::=$ | $\constDcl \baseTyp \var \exp$ | $\qquad \text{constants}$ |
+|   | $\vert$ | $\inst \baseTyp \args \var \stmts$ | $\qquad \text{instantiations (initialized)}$ |
+|   | $\vert$ | $\instNoBlock \baseTyp \args \var$ | $\qquad \text{instantiations (uninitialized)}$ |
+|   | $\vert$ | $\parserDcl \var \typVars {\overline {\prmDef_1}} {\overline {\prmDef_2}} \dcls \states$ | $\qquad \text{parsers}$ |
+|   | $\vert$ | $\controlDcl \var \typVars {\overline {\prmDef_1}} {\overline {\prmDef_2}} \dcls \stmts$ | $\qquad \text{controls}$ |
+|   | $\vert$ | $\funcDcl \baseTyp \var \typVars \prmDefs \stmts$ | $\qquad \text{functions}$ |
+|   | $\vert$ | $\externFuncDcl \baseTyp \var \typVars \prmDefs$ | $\qquad \text{extern functions}$ |
+|   | $\vert$ | $\varDclInit \baseTyp \var \exp$ | $\qquad \text{local variables (initialized)}$ |
+|   | $\vert$ | $\varDcl \baseTyp \var$ | $\qquad \text{local variables (uninitialized)}$ |
+|   | $\vert$ | $\valSetDcl \baseTyp \exp \var$ | $\qquad \text{parser value sets}$ |
```

Comment:
- [ ] disucss in meeting.
This might be a bit more general than P4's definition but not by much. Here's P4's definition:

```
valueSetDeclaration
  : optAnnotations
      VALUESET '<' baseType '>' '(' expression ')' name ';'
  | optAnnotations
      VALUESET '<' tupleType '>' '(' expression ')' name ';'
  | optAnnotations
      VALUESET '<' typeName '>' '(' expression ')' name ';'
  ;
```